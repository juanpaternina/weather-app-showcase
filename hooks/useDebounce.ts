import { useRef } from 'react';

const useDebounce = (callback: Function, delay: number) => {
  const timeoutRef = useRef<NodeJS.Timeout | null>(null);

  const debouncedFunction = (...args: any[]) => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
    }
    timeoutRef.current = setTimeout(() => {
      callback(...args);
    }, delay);
  };

  return debouncedFunction;
};

export default useDebounce;
