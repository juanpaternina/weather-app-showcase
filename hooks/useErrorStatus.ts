import { useCallback, useState } from 'react';

export const useErrorStatus = () => {
  const [error, setError] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');

  const updateErrorState = useCallback(
    (isError: boolean, errorMessage?: string) => {
      if (isError && errorMessage) {
        setErrorMessage(errorMessage);
        setError(isError);
        return;
      }

      if (isError && !errorMessage) {
        setErrorMessage('Something went wrong');
      }

      setError(isError);
      setErrorMessage('');
    },
    [],
  );

  return { error, errorMessage, updateErrorState };
};
