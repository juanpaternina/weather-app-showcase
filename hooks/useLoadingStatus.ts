import { useCallback, useState } from 'react';

export const useLoadingStatus = () => {
  const [loading, setLoading] = useState(false);

  const updateLoadingState = useCallback((isLoading: boolean) => {
    setLoading(isLoading);
  }, []);

  return { loading, updateLoadingState };
};
