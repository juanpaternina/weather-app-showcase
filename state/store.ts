import { configureStore } from '@reduxjs/toolkit';
import weatherSlice from './slices/weather';
import themeSlice from './slices/theme';
import { useDispatch, useSelector } from 'react-redux';

export const store = configureStore({
  reducer: {
    weather: weatherSlice,
    theme: themeSlice,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();

export default store;
