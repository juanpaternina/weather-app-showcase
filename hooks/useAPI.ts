import { useState } from 'react';
import { useErrorStatus } from './useErrorStatus';
import { useLoadingStatus } from './useLoadingStatus';

export const useAPI = <T>() => {
  const { loading, updateLoadingState } = useLoadingStatus();
  const { error, errorMessage, updateErrorState } = useErrorStatus();

  const [data, setData] = useState<T | null>(null);

  const fetchAPI = (request: string) => {
    updateLoadingState(true);
    fetch(request)
      .then((response) => response.json())
      .then((data) => {
        setData(data);
      })
      .catch(() => {
        updateErrorState(true, "Couldn't fetch data, please try again later");
        setData(null);
      })
      .finally(() => {
        updateLoadingState(false);
      });
  };

  return {
    data,
    error,
    errorMessage,
    loading,
    fetchAPI,
  };
};
