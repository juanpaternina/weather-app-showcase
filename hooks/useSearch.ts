import { useState } from 'react';
import { useLoadingStatus } from './useLoadingStatus';
import { useErrorStatus } from './useErrorStatus';
import { requestCreator } from '@/helpers/requestCreator';
import { City } from '@/types/City';
import { API_TYPES } from '@/constants/Api';

export const useSearch = () => {
  const { loading, updateLoadingState } = useLoadingStatus();
  const { error, errorMessage, updateErrorState } = useErrorStatus();

  const [data, setData] = useState<City[]>([]);

  const searchCity = (queryString: string): City[] => {
    updateLoadingState(true);

    fetch(
      requestCreator(API_TYPES.SEARCH, {
        q: queryString,
      }),
    )
      .then((response) => response.json())
      .then((data) => {
        console.log(data);
        setData(data);
      })
      .catch(() => {
        updateErrorState(true, "Couldn't fetch data, please try again later");
        setData([]);
      })
      .finally(() => {
        updateLoadingState(false);
      });
    return data;
  };

  return {
    data,
    error,
    errorMessage,
    loading,
    updateLoadingState,
    searchCity,
  };
};
