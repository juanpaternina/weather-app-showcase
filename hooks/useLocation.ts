import { useCallback } from 'react';
import * as Location from 'expo-location';

import { useAppDispatch } from '@/state/store';
import { updateUserLocation } from '@/state/slices/weather';

export const useLocation = () => {
  const dispatch = useAppDispatch();

  const getLocation = useCallback(async () => {
    const { status } = await Location.requestForegroundPermissionsAsync();
    if (status !== 'granted') {
      return;
    }

    const { coords } = await Location.getCurrentPositionAsync({});

    dispatch(
      updateUserLocation({ lat: coords.latitude, lng: coords.longitude }),
    );
  }, [dispatch]);

  return { getLocation };
};
