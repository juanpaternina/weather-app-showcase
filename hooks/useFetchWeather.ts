import { requestCreator } from '@/helpers/requestCreator';
import { API_TYPES } from '@/constants/Api';
import { useAPI } from './useAPI';
import { WeatherAPI } from '@/types/api/weather';

export const useFetchWeather = () => {
  const { data, error, errorMessage, loading, fetchAPI } = useAPI<WeatherAPI>();

  const fetchWeather = (lat: number, long: number) => {
    const requestURL = requestCreator(API_TYPES.FORECAST, {
      q: `${lat},${long}`,
      days: 5,
    });
    fetchAPI(requestURL);
  };

  return {
    data,
    error,
    errorMessage,
    loading,
    fetchWeather,
  };
};
