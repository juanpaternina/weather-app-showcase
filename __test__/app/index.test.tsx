import React from 'react';
import { render } from '@testing-library/react-native';
import App from '@/app/index';

import { useLocation } from '@/hooks/useLocation';

import { useAppDispatch, useAppSelector } from '@/state/hooks';
import { useFetchWeather } from '@/hooks/useFetchWeather';

import data from '@/mocks/weather.json';

// Mock components
jest.mock('@/components/Header', () => ({
  Header: jest.fn(() => <></>),
}));
jest.mock('@/components/Date', () => ({
  Date: jest.fn(() => <></>),
}));
jest.mock('@/components/Degrees', () => ({
  Degrees: jest.fn(() => <></>),
}));
jest.mock('@/components/Condition', () => ({
  Condition: jest.fn(() => <></>),
}));
jest.mock('@/components/Summary', () => ({
  Summary: jest.fn(() => <></>),
}));
jest.mock('@/components/ForecastDay', () => ({
  ForecastDay: jest.fn(() => <></>),
}));

jest.mock('@/state/hooks', () => ({
  useAppDispatch: jest.fn(),
  useAppSelector: jest.fn(),
}));

jest.mock('@/hooks/useLocation', () => ({
  useLocation: jest.fn(),
}));

jest.mock('@/hooks/useFetchWeather', () => ({
  useFetchWeather: jest.fn(),
}));

const mockDispatch = jest.fn();

describe('App Component', () => {
  beforeEach(() => {
    const state = {
      weather: {
        location: {
          lat: 51.509865,
          lng: -0.118092,
        },
      },
      theme: {
        color: 'white',
      },
    };

    const getLocationMock = jest.fn();

    (useAppDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    (useAppSelector as unknown as jest.Mock).mockImplementation((f) =>
      f(state),
    );

    (useLocation as jest.Mock).mockReturnValue({
      getLocation: getLocationMock,
    });
    jest.clearAllMocks();
  });

  test('renders correctly', () => {
    (useFetchWeather as jest.Mock).mockReturnValue({
      data: data,
      error: false,
      errorMessage: '',
      loading: false,
      fetchWeather: jest.fn(),
    });

    const { toJSON } = render(<App />);

    expect(toJSON()).toMatchSnapshot();
  });

  test("Should show an error message if there's an error", () => {
    (useFetchWeather as jest.Mock).mockReturnValue({
      data: null,
      error: true,
      errorMessage: 'Error message',
      loading: false,
      fetchWeather: jest.fn(),
    });

    const { toJSON } = render(<App />);

    expect(toJSON()).toMatchSnapshot();
  });
});
