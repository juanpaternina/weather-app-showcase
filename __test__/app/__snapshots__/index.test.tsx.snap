// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Should show an error message if there's an error 1`] = `
<Text
  style={
    {
      "color": "black",
      "fontFamily": "medium",
      "fontSize": 20,
      "fontWeight": "bold",
    }
  }
>
  Error message
</Text>
`;

exports[`App Component renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
      "padding": 0,
    }
  }
>
  <View
    style={
      {
        "marginTop": 4,
      }
    }
  />
</RCTSafeAreaView>
`;
